src/sanity_side.rs

```diff
@@ -87,284 +106,250 @@ impl Side for SideImpl {
-        let mut local_qp_map = QP_MAP.write().unwrap();
-        local_qp_map.push(qp.clone());
-
-        let mut resp = CreateQpResponse::default();
-        resp.set_qp_id((local_qp_map.len() - 1).cast());
-        resp.set_qp_num(qp_num);
-
-        let f = sink.success(resp).map_err(|_| {}).map(|_| ());
-        ctx.spawn(f);
-    }
-
-    fn connect_qp(
+    #[tokio::main]
+    async fn connect_qp(
         &mut self,
         ctx: grpcio::RpcContext,
         req: proto::message::ConnectQpRequest,
         sink: grpcio::UnarySink<proto::message::ConnectQpResponse>,
     ) {
-        let ib_port = req.get_ib_port_num();
-        let local_qp_map = QP_MAP.read().unwrap();
-        let qp = local_qp_map.get(req.get_qp_id().cast::<usize>()).unwrap();
-        let flag = req.get_access_flag();
-        let gid_idx = req.get_gid_idx();
-        let remote_qp_num = req.get_remote_qp_num();
-        let remote_lid = req.get_remote_lid();
-        let remote_gid: [u8; 16] = req.get_remote_gid().try_into().unwrap();
-        let timeout = req.get_timeout();
-        let retry_cnt = req.get_retry();
-        let rnr_retry = req.get_rnr_retry();
-        let mtu = req.get_mtu();
-        let sq_start_psn = req.get_sq_start_psn();
-        let rq_start_psn = req.get_rq_start_psn();
-        let max_rd_atomic = req.get_max_rd_atomic();
-        let max_dest_rd_atomic = req.get_max_dest_rd_atomic();
-        let min_rnr_timer = req.get_min_rnr_timer();
-
-        rdma::ibv::modify_qp_to_init(ib_port.cast(), *qp, rdma_sys::ibv_access_flags(flag.cast()));
-        rdma::ibv::modify_qp_to_rtr(
-            gid_idx.cast(),
-            ib_port.cast(),
-            *qp,
-            remote_qp_num,
-            remote_lid.cast(),
-            u128::from_be_bytes(remote_gid),
-            mtu,
-            rq_start_psn,
-            max_dest_rd_atomic.cast(),
-            min_rnr_timer.cast(),
-        );
-        debug!(
-            "Transfer to RTS, qp = {:?}, timeout = {}, retry_cnt = {}, rnr_retry = {}",
-            (*qp).inner,
-            timeout,
-            retry_cnt,
-            rnr_retry
-        );
-        rdma::ibv::modify_qp_to_rts(
-            *qp,
-            timeout.cast(),
-            retry_cnt.cast(),
-            rnr_retry.cast(),
-            sq_start_psn,
-            max_rd_atomic.cast(),
-        );
+        let mut rdma_map = RDMA_MAP.write().await;
+        let rdma = rdma_map.get_mut(req.get_dev_name()).unwrap();
+
+        let remote = QueuePairEndpointBuilder::default()
+            .qp_num(req.get_remote_qp_num())
+            .lid(req.get_remote_lid().cast())
+            .gid(Gid::from_raw(req.get_remote_gid().try_into().unwrap()))
+            .build()
+            .unwrap();
+
+        rdma.ibv_connect(remote).await.unwrap();
 
         let resp = ConnectQpResponse::default();
         let f = sink.success(resp).map_err(|_| {}).map(|_| ());
         ctx.spawn(f)
     }
 
-    fn remote_read(
+    #[tokio::main]
```

Comment:
It'll create a runtime for each call. Is that efficient?

Reply:
That is not efficient and just for the convenience of coding.
We can also use only one staic `RUNTIME` in every call, such as:
```
fn remote_read(){
  RUNTIME.block_on/spawn(
    async {...}
  )
}
```
Considering that this is a project for testing, convenience is given priority so I did this. 
What is your advise or do you have any other ideas.

Reply:
Agree. Let's change it when performance is too bad.